cmake/multi_image.cmake

```
@@ -0,0 +1,159 @@
+#
+# Copyright (c) 2019 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+#
+
+function(image_board_selection board_in board_out)
+  # Unless parent was "ns" in which case we assume that the child images are
+  # all secure, and should be build using the secure version of the board.
+  # It is assumed that only the root app will be built as non-secure.
+  # This is not a valid assumption as there might be multiple non-secure
+  # images defined. With this technique, it is not possible to have multiple
+  # images defined as non-secure.
+  # TODO: Allow multiple non-secure images by using Kconfig to set the
+  # secure/non-secure property rather than using a separate board definition.
+  set(${board_out} ${board_in} PARENT_SCOPE)
+  if(${board_in} STREQUAL nrf9160_pca10090ns)
+    set(${board_out} nrf9160_pca10090 PARENT_SCOPE)
+    message("Changed board to secure nrf9160_pca10090 (NOT NS)")
+  endif()
+
+  if(${BOARD} STREQUAL nrf9160_pca20035ns)
+    set(${board_out} nrf9160_pca20035 PARENT_SCOPE)
+    message("Changed board to secure nrf9160_pca20035 (NOT NS)")
+  endif()
+
+  if(${BOARD} STREQUAL actinius_icarus_ns)
+    set(BOARD actinius_icarus)
+    message("Changed board to secure actinius_icarus (NOT NS)")
+  endif()
+endfunction()
+
+function(image_menuconfig target_out command_out)
+  if(EXTRA_KCONFIG_TARGETS)
+    set(${target_out}  "-DEXTRA_KCONFIG_TARGETS=${EXTRA_KCONFIG_TARGETS}" PARENT_SCOPE)
+    set(${command_out} "-DEXTRA_KCONFIG_TARGET_COMMAND_FOR_${EXTRA_KCONFIG_TARGETS}=${EXTRA_KCONFIG_TARGET_COMMAND_FOR_${EXTRA_KCONFIG_TARGETS}}" PARENT_SCOPE)
+  endif()
+endfunction()
+
+function(zephyr_add_external_image name sourcedir)
+  string(TOUPPER ${name} UPNAME)
+
+  if (CONFIG_${UPNAME}_BUILD_STRATEGY_USE_HEX_FILE)
+    assert_exists(CONFIG_${UPNAME}_HEX_FILE)
+    message("Using ${CONFIG_${UPNAME}_HEX_FILE} instead of building ${name}")
+  elseif (CONFIG_${UPNAME}_BUILD_STRATEGY_SKIP_BUILD)
+    message("Skipping building of ${name}")
+  else()
+    # Build normally
+    zephyr_add_external_image_from_source(${name} ${sourcedir})
+  endif()
+endfunction()
+
+function(zephyr_add_external_image_from_source name sourcedir)
+  string(TOUPPER ${name} NAME)
+  set(${NAME}_CMAKE_BINARY_DIR ${CMAKE_BINARY_DIR}/${name})
+  set(${NAME}_PROJECT_BINARY_DIR ${CMAKE_BINARY_DIR}/${name}/zephyr)
+
+  get_cmake_property(VARIABLES_CACHED CACHE_VARIABLES)
+  get_cmake_property(VARIABLES              VARIABLES)
+
+  unset(image_cmake_args)
+  foreach(var_name
+      ${VARIABLES}
+      ${VARIABLES_CACHED}
+      )
+    if("${var_name}" MATCHES "^${name}_(.+)") # TODO: How expensive?
```

```suggestion
    if("${var_name}" MATCHES "^${NAME}_(.+)") # TODO: How expensive?
```
alternatively:
```suggestion
    if("${var_name}" MATCHES "^${UPNAME}_(.+)") # TODO: How expensive?
```
if other change is taken in.
As this allows users to write: `cmake -DSPM_<ARG>=<val>` instead of mixing cases as: `cmake -Dspm_<ARG>=<val>`

---

No, we want to mix case to reduce risk of aliasing issues and to stay backwards-compatible,

---

ok, fine.
Btw. is this documented anywhere ?
As this is old feature I suppose there is some documentation somewhere that describes the existing multi image use-case and handling of conf settings for the subimage.
Including, how do the user know the name of the sub-image, e.g. `spm`, `mcuboot`, `b0` ?

---

Found something: http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/ug_multi_image.html?highlight=multi%20image#multi-image-builds
but nothing regarding the configuration of sub-image and how to know the name of the sub image.